types/react/test/hooks.tsx

```diff
@@ -0,0 +1,120 @@
+import * as React from "react";
+
+interface PersonProps {
+    name: string;
+    age: number;
+}
+export function Person(props: PersonProps) {
+    return <div>
+        hello! I'm {props.name} and I'm {props.age} years old!
+    </div>;
+}
+
+export interface FancyButtonProps {
+    onClick: () => void;
+    children?: React.ReactNode;
+}
+export interface FancyButton {
+    fancyClick(): void;
+}
+export const FancyButton = React.forwardRef((props: FancyButtonProps, ref: React.Ref<FancyButton>) => {
+    const buttonRef = React.useRef<HTMLButtonElement | null>(null);
+    const [count, setCount] = React.useState(0);
+
+    React.useImperativeMethods(ref, () => ({
+        fancyClick() {
+            buttonRef.current!; // $ExpectType HTMLButtonElement
+        },
+        getClickCount() {
+            return count;
+        }
+    }));
+
+    return <button onClick={() => { setCount(count + 1); props.onClick(); }}>
+        {props.children}
+    </button>;
+});
+
+interface AppState {
+    name: string;
+    age: number;
+}
+
+type AppActions =
+    | { type: "getOlder" }
+    | { type: "resetAge" };
+
+function reducer(s: AppState, action: AppActions): AppState {
+    switch (action.type) {
+        case "getOlder":
+            return { ...s, age: s.age + 1 };
+        case "resetAge":
+            return { ...s, age: 0 };
+    }
+}
+
+const initialState = {
+    name: "Daniel",
+    age: 26
+};
+
+export function App() {
+    const [state, dispatch] = React.useReducer(reducer, initialState);
+    const birthdayRef = React.useRef<FancyButton | null>(null);
+
+    React.useLayoutEffect(() => {
+        birthdayRef.current!.fancyClick();
+    });
+
+    return <>
+        <Person {...state} />
+        <FancyButton onClick={() => dispatch({ type: "getOlder" })}>
+            Birthday time!
+        </FancyButton>
+        <FancyButton onClick={() => dispatch({ type: "resetAge" })}>
+            Let's start over.
+        </FancyButton>
+    </>;
+}
+
+interface Context {
+    test: true;
+}
+const context = React.createContext<Context>({ test: true });
+
+function useEveryHook(ref: React.Ref<{ id: number }>|undefined): () => boolean {
+    const value: Context = React.useContext(context);
+    const [, setState] = React.useState(() => 0);
+    const [reducerState, dispatch] = React.useReducer(reducer, initialState, { type: 'resetAge' });
+    const didLayout = React.useRef(false);
+
+    const id = React.useMemo(() => Math.random(), []);
+    React.useImperativeMethods(ref, () => ({ id }), [id]);
```

Comment:
We're actually hitting against TypeScript's inability to declare the input's variance (the first argument is nonvariant). The callback here is allowed to return empty object, causing `T` to infer as `{}`, and that wouldn't be a type error(!) because `ref` would still be assignable to `Ref<{}>|undefined` under covariance rules.

---

PS: I've edited this comment countless times because I still don't understand covariance/contravariance intuitively 😩

Reply:
Found a hackish solution 😆